Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInWarningMatcher.kt

```diff
@@ -0,0 +1,27 @@
+package de.rki.coronawarnapp.presencetracing.risk
+
+import de.rki.coronawarnapp.eventregistration.checkins.CheckInRepository
+import de.rki.coronawarnapp.eventregistration.checkins.download.TraceTimeIntervalWarningRepository
+import de.rki.coronawarnapp.eventregistration.checkins.split.splitByMidnightUTC
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.flow.firstOrNull
+import javax.inject.Inject
+
+class CheckInWarningMatcher @Inject constructor(
+    private val checkInsRepository: CheckInRepository,
+    private val traceTimeIntervalWarningRepository: TraceTimeIntervalWarningRepository,
+    private val dispatcherProvider: DispatcherProvider
+) {
+    suspend fun execute(): List<CheckInWarningOverlap> {
+        val checkIns = checkInsRepository.allCheckIns.firstOrNull() ?: return emptyList()
+
+        val warningPackages = traceTimeIntervalWarningRepository.allWarningPackages.firstOrNull() ?: return emptyList()
+
+        val splitCheckIns = checkIns.flatMap { it.splitByMidnightUTC() }
+
+        return launchMatching(splitCheckIns, warningPackages, dispatcherProvider.IO)
```

Comment:
Love the parallelization effort!

We might have to cap the parallel executions though.
Currently we launch one coroutine (possibly one thread due to `IO` dispatcher) per `warningPackage`, which may be 24*15, though capped at 64 by kotlin (afaik).  Parallel I/O (reading the packages) is usually the bottle neck on most devices, especially low end ones, so 64 parallel attempts to read the files may starve the system of I/O resources and could cause the device to lag during our background execution.

This is all theoretical though, we'd need to test this with an actual set of warning packages (@mlenkeit is something like that available?).

Maybe we should cap parallel execution at 2-4 for now? What do you think? Anything above 4 might have diminishing returns anyways :thinking: